### ISSUE TITLE
`end` is slow when used to index trailing dimensions of multidimensional arrays

### ISSUE THREAD

**Issue by colintbowers** [JuliaLang/julia#19389]
If `end` is used to index, for example, the trailing dimension of a matrix when double looping over the matrix, it is roughly 3 to 5 times slower than pre-allocating the number of columns and using that integer in place of `end`. See, for example, [here](http://stackoverflow.com/questions/40744675/efficient-element-wise-matrix-operations-in-julia?noredirect=1#comment68732021_40744675).

It appears to stem from a call to `Base.trailingsize` that LLVM is not able to inline.

I'm not sure if this can be magically fixed, but if not, then possibly it should be mentioned in the performance tips section of the docs.

**Comment by mbauman** [comment]
I think at the least we can add a fast-path in the common case where `n == ndims(A)`. LLVM will elide that branch if it gets inlined with constant `n`.

**Comment by KristofferC** [comment]
This is equally fast now whether using `end` or storing it in a variable.

```jl
function convolve(M::Array{Complex{Float64},2}, K::Array{Float64,2}, p::Int)
    (n,m) = size(M)
    res = zeros(Complex{Float64},n)
    for k=1:p
        for l=1:n
            res[l] += M[l,k]*K[l,end-p+k]
        end
    end
    return res
end

function convolve2(M::Array{Complex{Float64},2}, K::Array{Float64,2}, p::Int)
    (n,m) = size(M)
    z = size(K,2)
    res = zeros(Complex{Float64},n)
    for k=1:p
        for l=1:n
            res[l] += M[l,k]*K[l,z-p+k]
        end
    end
    return res
end

M=complex(rand(2000,2000))
K=rand(2000,2000)
@btime convolve(M,K,2000)
@btime convolve2(M,K,2000)
```

```jl
julia> @btime convolve(M,K,2000)
  5.510 ms (2 allocations: 31.33 KiB)

julia> @btime convolve2(M,K,2000)
  5.500 ms (2 allocations: 31.33 KiB)
```

**Comment by mbauman** [comment]
(Fixed by #21750)
